### DISCUSSION TITLE
Protocol with arbitrary singature matching?

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#6049] [Q&A] [answered]
I'm trying to define a protocol for [PyTorch's `Dataset` class ](https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset). It only requires the definition of `__init__`, `__len__`, and `__getitem__` without any restrictions as for the parameters. I am having trouble finding a solution that appeases both pyright and mypy.

I would like your opinion on the 2 solutions presented below, and if pyright should support both of them. I think pyright's error in function `bar` is a little unexpected.

```python
from typing import Any, Callable, Protocol

# pyright is fine with this and everything works as expected. I think this looks best, also.
class Dataset(Protocol):
    __init__: Callable[..., None]
    __len__: Callable[..., int]
    __getitem__: Callable[..., Any]

# Courtesy @gvanrossum: https://github.com/python/mypy/issues/9560#issuecomment-705955103.
# This is the solution I have seen people advising for mypy notes/error for the above solution. IMO it really does look horrible.
class HorribleDataset(Protocol):
    __init__: Callable[..., None]

    @property
    def __len__(self) -> Callable[..., int]:
        ...

    @property
    def __getitem__(self) -> Callable[..., Any]:  # pylint: disable=unexpected-special-method-signature
        ...

# an arbitrary implementation of the Protocols
class ExampleDataset:
    def __init__(self, root: str) -> None:
        ...

    def __len__(self) -> int:
        return 1

    def __getitem__(self, idx: int) -> Any:
        ...

#########################################################################################################

def foo(dataset: Dataset) -> None:
    print(len(dataset))  # all OK

# pyright : OK
# mypy(error) : Argument 1 to "foo" has incompatible type "ExampleDataset"; expected "Dataset"  [arg-type]
# mypy(note) : Protocol member Dataset.__getitem__ expected settable variable, got read-only attribute
# mypy(note) : Protocol member Dataset.__len__ expected settable variable, got read-only attribute
foo(ExampleDataset("/data/"))

def bar(dataset: HorribleDataset) -> None:
    print(len(dataset))  # pyright : Type "property" cannot be assigned to type "(self: Dataset) -> int"

# all OK
bar(ExampleDataset("/data/"))

```

**Comment by erictraut** [comment, marked as answer]
If you use the syntax `__init__: Callable[..., None]`, that implies that it's an instance variable. Probably not what you intend.

Is your intent that these three methods should accept any set of parameters? That doesn't sound correct to me, at least not for `__len__` and `__getitem__` because those are magic methods defined by the language, and they must have specific well-defined signatures. If they have some other signature, they won't work with the runtime, so it doesn't make sense to allow an arbitrary signature in these cases.

As for `__init__`, all classes inherit an `__init__` from `object`, so every if your intent is to say "this class or its superclass must supply _some_ `__init__` method with an arbitrary signature", then every class will satisfy that requirement even if you don't supply an `__init__` in the protocol. So I recommend removing that from the protocol. It's not doing any good anyway.

So here's my recommendation:
```python
class Dataset(Protocol):
    def __len__(self) -> int:
        ...

    def __getitem__(self, key: Any, /) -> Any:
        ...
```

This works with both pyright and mypy.

**Reply by ringohoffman** [reply]
Thanks! What is your recommendation for non-magic methods?

**Reply by erictraut** [reply]
Can you elaborate? I think I need more details about your requirements and use case.

Keep in mind that protocols are meant to define the expected "shape" of the interface. This guarantees compatibility between a method's caller and a callee. If you don't know the expected signature for a method, then it doesn't make sense to put in a protocol definition.
